all-sdk/external-client-configuration.md

```diff
@@ -0,0 +1,526 @@
+    * Because a user with a profile configured with a codec should expect that codec to be used, not silently ignored.
+
+Simplest example:
+
+```typescript
+const options = loadClientOptionsFromConfig();
+const connection = Connection.connect(options);
+const client = new Client({ connection, ...options });
+
+// Or maybe:
+
+const { connectionOptions, clientOptions } = loadClientOptionsFromConfig();
+const connection = Connection.connect(connectionOptions);
+const client = new Client({ connection, ...clientOptions });
+```
+
+#### Python SDK Idea
+
+* Either new `temporalio.contrib.envconfig` module and "extra" for the TOML library, or new
+  `temporalio.client.envconfig` module that uses Rust Core.
+  * ❓ Should we use Rust Core for the toml/env loading?
+* New `ClientConfig` dataclass (and others as needed).
+* Static method `ClientConfig.load` with kwarg parameters of `config_file: Optional[str] = None`,
+  `disable_file: bool = False`, `disable_env: bool = False` and it returns a `ClientConfig`.
+* New `TypedDict` in `client` for `ConnectConfig` that matches all params in `connect`.
+* Static method `ClientConfig.load_client_connect_config` with positional parameter of
+  `profile: str = 'default'` and kwarg parameters of `config_file: Optional[str] = None`, `disable_file: bool = False`,
+  and `disable_env: bool = False` and it returns a `client.ConnectConfig`
+  * 💭 Why not call it `load_client_config`?
+    * Because this class is a connect config and that is confusing.
+* Instance method `to_client_connect_config` on the `ClientConfig`.
+* Like TypeScript, Python SDK _does not_ support codec settings and will error if seen in config.
+
+Simplest example:
+
+```python
+config = envconfig.ClientConfig.load_client_connect_config()
+client = await Client.connect(**config)
+```
+
+#### .NET SDK Idea
+
+* Either new `Temporalio.Extensions.EnvConfig` project with TOML dependency, or new `Temporalio.Client.EnvConfig`
+  namespace that uses Rust Core.
+  * ❓ Should we use Rust Core for the toml/env loading?
+* New `ClientConfig` record (and others as needed).
+* Static method `ClientConfig.Load(string? configFile = null, bool disableFile = false, bool disableEnv = false)` that
+  returns `ClientConfig`.
+* Static method `ClientConfig.LoadClientConnectOptions` with same params.
+* Instance method on `ClientConfig` for `ToClientConnectOptions`.
+* Like TypeScript, .NET SDK _does not_ support codec settings and will error if seen in config.
+
+Simplest example:
+
+```csharp
+var options = ClientConfig.LoadClientConnectOptions();
+var client = await TemporalClient.ConnectAsync(options);
+```
+
+### Samples
+
+* All samples will be updated to load from config and default to local dev server w/ default namespace.
+  * In samples we try to avoid shared code. In .NET and Python, there is no default target host, and in Ruby there will
+    not even be a default namespace. So every sample in these languages is going to have to not only have the
```

Comment:
> So every sample in these languages is going to have to not only have the load-from-config code, but also conditionally apply the target host as `localhost:7233` if not set in external config.

🤔 That kind of defeats on of the purpose of the load-from-config utility function, and would have little demonstrative value in samples as most "sample-level" users would, in practice, either fully adopt the config file approach, or fully specify their client config programmatically.

I think we should revisit the question on either load-from-config should fill in a default address and namespace. I totally understand why that's tricky in these languages, but I'm not sure the current proposal is the most reasonable compromise.

Reply:
> I think we should revisit the question on either load-from-config should fill in a default address and namespace.

There is no default address and namespace in some SDKs is the point. We have _removed_ this default as we've become more mature knowing that the older defaults (which we cannot change) are not as useful in deployed code. But for those SDKs w/out a default address/namespace there _is_ a default address/namespace in samples, so they must set it. But that's only for samples, we shouldn't affect non-sample code w/ sample-only defaults.

Reply:
Why not just make it required for the config to specify these things (even for langs where the in-code options allow a default)? It sure would make the config file more readable by default, and can still be overridden with env var.

Reply:
Because some people just want to configure, say, the remote codec with a config file or env var. Or there is nothing in config and only address is set by env var and namespace is expected to be set by user. Or config file only sets address and namespace is set via env var. All config keys are optional (as is even the presence of the config file). It's only when the client options are used does whether a value being present matter.